util/ranger/ranger.go

```diff
@@ -88,9 +90,12 @@ func points2Ranges(sctx sessionctx.Context, rangePoints []*point, tp *types.Fiel
 			HighExclude: endPoint.excl,
 			Collators:   []collate.Collator{collate.GetCollator(tp.GetCollate())},
 		}
+		if len(ranges) == 0 && rangeMaxSize > 0 && ran.MemUsage()*int64(len(rangePoints))/2 > rangeMaxSize {
```

Comment:
Should we handle part of `rangePoints` in the next PR?

Reply:
No, we don't restrict the mem usage of `rangePoints` for simplicity. For single column, building ranges procedure is like `conditions -> rangePoints -> ranges`. The mem usage of `conditions`, `rangePoints` and `ranges` is almost linear. In other words, if you want to construct single-column `ranges` consuming lots of memory, the corresponding `conditions` must contain lots of constants, which is probably not feasible due to sql length limit or other restrictions. However, for multi column index, several IN list conditions can easily build ranges consuming lots of memory(the number of ranges is the product of IN list lengths, see #30755). Actually, if we only want to prevent OOM caused by building ranges, restricting the mem usage of building multi column index ranges may just be ok. However, for consistency, we restrict the mem usage of building all kinds of ranges(table range, single-column index range, multi column index range...), which is also the behavior of MySQL. And we build all of `rangePoints` without mem limit for single column for simplicity.